---
slug: '/notes/Use-CollectHsMetrics-(Picard)-on-GemBS-Bam-file'
date: '2022-02-09'
title: 'Use CollectHsMetrics (Picard) on GemBS Bam file'
tags: ['Debug', 'BAM', 'GemBS']
abstract: 'I spend two days to make it work, running CollectHsMetrics on GemBS created BAM. The bug is created by different ways of MD5 generation. Here is a record how to make it work.'
---



The task here is to run Picard tool [CollectHsMetrics](https://gatk.broadinstitute.org/hc/en-us/articles/360036856051-CollectHsMetrics-Picard-), initially it’s not working, after long time debug, I eventually found the error was caused by different MD5 in head tag. 

## 1. Prepare dict file for reference with CreateSequenceDictionary (Picard)
To run CollectHsMetrics, the first step is to create a reference.dict for the reference used for fastq-2-bam alignment. The code is below:

```
java -jar picard.jar CreateSequenceDictionary -R hg38.fa -O hg38.dict

```

After above command, a `hg38.dict` file will be created in the current folder. It looks like below:

```
tian@slms-ci-beck-02 /scratch1/Tian/Cansor/EMseq/13.PicardHSMetric $ head hg38.dict
@HD	VN:1.6
**@SQ	SN:chr1	LN:248956422	M5:2648ae1bacce4ec4b6cf337dcae37816	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz**
@SQ	SN:chr10	LN:133797422	M5:907112d17fcb73bcab1ed1c72b97ce68	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr11	LN:135086622	M5:1511375dc2dd1b633af8cf439ae90cec	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr11_KI270721v1_random	LN:100316	M5:9654b5d3f36845bb9d19a6dbd15d2f22	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr12	LN:133275309	M5:e81e16d3f44337034695a29b97708fce	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr13	LN:114364328	M5:17dab79b963ccd8e7377cef59a54fe1c	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr14	LN:107043718	M5:acbd9552c059d9b403e75ed26c1ce5bc	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr14_GL000009v2_random	LN:201709	M5:862f555045546733591ff7ab15bcecbe	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz
@SQ	SN:chr14_GL000225v1_random	LN:211173	M5:63945c3e6962f28ffd469719a747e73c	UR:file:/scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/hg38.fa.gz

```

Then use this file to create interval with Twist target regions.

```
java -jar picard.jar BedToIntervalList I=Covered_Targets_Twist_Methylome_V1_hg38.txt  O=Twist.interval_list SD=hg38.replaced.dict

```

However, after this step, the standard picard function is not working, after a long time search, I found the problem is MD5.

## 2. Problem happen: Different MD5 between methods

When I check the bam files’ head tag. The M5 is different:

```
tian@slms-ci-beck-02 /scratch1/Tian/Cansor/EMseq/2.GemBS_Analysis/Compare3Set/mapping/J080001 $ samtools view -H J080001.bam | head
@HD	VN:1.4	SO:coordinate
**@SQ	SN:chr1	LN:248956422	M5:a004bc1b0bf05fc668cab6bbfd93d3eb**
@SQ	SN:chr10	LN:133797422	M5:6b420cbb22daea77d7cc930c0a00f812
@SQ	SN:chr11	LN:135086622	M5:0d4e0be5c4e5bc0f12912894f21a5dd8
@SQ	SN:chr11_KI270721v1_random	LN:100316	M5:e3075f8f08d2f270b013932f76fbe247
@SQ	SN:chr12	LN:133275309	M5:e1507ba70028a65b3f5a81b594e6f0fe
@SQ	SN:chr13	LN:114364328	M5:7110500758388b169fe631b212b7e56c
@SQ	SN:chr14	LN:107043718	M5:f37e77fdbacb1a0f1be5e2bf25df343d
@SQ	SN:chr14_GL000009v2_random	LN:201709	M5:8facccfe9de958a6033a33c35ab4ebb2
@SQ	SN:chr14_GL000225v1_random	LN:211173	M5:08d9a7a0b240ed926ed17a32929b785d

```

The MD5 head of each bam file is generated by `hg38.gemBS.contig_md5` file after `gemBS index`

. **In other word, why gemBS and Picard created different MD5 for each contig?**

Inspired by [this post](https://github.com/samtools/samtools/issues/704), I tested a bit, after extract a short contig chrUn_KI270373v1 from hg38.fa. Then create a new `test.fa` as below:

```
>chrUn_KI270373v1
aaaactagacagaagcattctcagaagcgactttgtgatgtgtgcattct
actcccatagttgaacctttcttttgatagagcagtctggaaacactctt
tttgtcgaatctgcaaatggacttttggagcgctttgaaggctatggtgg
aaaaggaaatatcttcacataaaaactagacggaagcattctcagaaact
tctttgtgttgtgttcattcaactcacatagttgaactttacttttgata
gagcagtttggaagcccttttttattaagtctgcaggaggatatttggag
ctctttgaatccttcgttggaaacgggaatagcttcatataaatactaga
aagaagcattctcagaaaccactttgtgatgtgtgcattcaactcacaca
gttgaaccttccttttgatagagcagtttggaaacactctttttgtagaa
tgtgcatgtagatatttgtagcgaatagaagccttcgtttgaaataggaa
tatcttcacctaaaaactagatggaagcattgtcagaaagtcctttgtga
tgtgtgcattcaactcacagaactgaaccttccttcagatagagaagttt
tgaaacactctttttgtagaatctgcatttacatctttggagcgctttga
agtcttcgttggaaacgcaaataccttcacataaaaaatagacggagcca
ttctcagaaaattatttgtgatgtgtgcattcaactcacagagttgaacc
atctttttgatagagcagttttgaaacagtctttttgtagaatctggagg
cgcatatttagagcgctttgaagccttccttggaaatgggattatcttca
cataaaaactagacagctttctacatatggctagtcagttttcccagcac
catttattaaagagggaatcctttccccattgcttgtttttctcaggttt
gtcaaagatcagatagttgtagatatgcggcgttatttctgagggctctg
ttctgttccgttgatctatatctctgttttggtaccagtaccatgctgtt
ttggttactggaaccttgtagtatagtttgaagtcaggtagtgtgatgcc
tccagctttgttcttttggcttaggattgacttggtgatgcgggctcttt
tttggttccatatgaactttaaagtagcttttccaattctgtgaagaaag
gcattggtagcttgatggggatggcactgaatctgcaaattaccttgggc
agtatggccattttcacgatattgattcttcctacccatgagcgtgggaa
tgttcttccatttgtttgtatcctcttttatttccttgagtagttggttt
gtagttctctttgaaagaggtccttcacatccctttgtaggttggattcc
taggtattttattctctttgaagcaattgtgaatggggagttcactcatg
a

```

Then I manually created the MD5 by reading the string into R.

```
> contig <- read.table("test.fa", skip=1)
> contig <- paste0(contig[,1], collapse="")
> contig
[1] "aaaactagacagaagcattctcagaagcgactttgtgatgtgtgcattctactcccatagttgaacctttcttttgatagagcagtctggaaacactctttttgtcgaatctgcaaatggacttttggagcgctttgaaggctatggtggaaaaggaaatatcttcacataaaaactagacggaagcattctcagaaacttctttgtgttgtgttcattcaactcacatagttgaactttacttttgatagagcagtttggaagcccttttttattaagtctgcaggaggatatttggagctctttgaatccttcgttggaaacgggaatagcttcatataaatactagaaagaagcattctcagaaaccactttgtgatgtgtgcattcaactcacacagttgaaccttccttttgatagagcagtttggaaacactctttttgtagaatgtgcatgtagatatttgtagcgaatagaagccttcgtttgaaataggaatatcttcacctaaaaactagatggaagcattgtcagaaagtcctttgtgatgtgtgcattcaactcacagaactgaaccttccttcagatagagaagttttgaaacactctttttgtagaatctgcatttacatctttggagcgctttgaagtcttcgttggaaacgcaaataccttcacataaaaaatagacggagccattctcagaaaattatttgtgatgtgtgcattcaactcacagagttgaaccatctttttgatagagcagttttgaaacagtctttttgtagaatctggaggcgcatatttagagcgctttgaagccttccttggaaatgggattatcttcacataaaaactagacagctttctacatatggctagtcagttttcccagcaccatttattaaagagggaatcctttccccattgcttgtttttctcaggtttgtcaaagatcagatagttgtagatatgcggcgttatttctgagggctctgttctgttccgttgatctatatctctgttttggtaccagtaccatgctgttttggttactggaaccttgtagtatagtttgaagtcaggtagtgtgatgcctccagctttgttcttttggcttaggattgacttggtgatgcgggctcttttttggttccatatgaactttaaagtagcttttccaattctgtgaagaaaggcattggtagcttgatggggatggcactgaatctgcaaattaccttgggcagtatggccattttcacgatattgattcttcctacccatgagcgtgggaatgttcttccatttgtttgtatcctcttttatttccttgagtagttggtttgtagttctctttgaaagaggtccttcacatccctttgtaggttggattcctaggtattttattctctttgaagcaattgtgaatggggagttcactcatga"
>
> library(digest)
> digest(contig, algo="md5", serialize=F)
[1] "8abd598da8e385b0f3c48f0b6bc9e35e"
>
> digest(toupper(contig), algo="md5", serialize=F) # Picard value
[1] "b174fe53be245a840cd6324e39b88ced"
>

```

The MD5 value `8abd598da8e385b0f3c48f0b6bc9e35e` is exactly the same as the value in `hg38.gemBS.contig_md5`  (below line with chrUn_KI270373v1 start).

```
# hg38.gemBS.contig_md5 file
chrUn_KI270376v1    LN:1136 M5:afa726dc6156278c2cda16e84f532850
chrUn_KI270374v1    LN:2656 M5:a404c1bedb4862c64d493643b3cf786f
chrUn_KI270372v1    LN:1650 M5:2fa607121376431d2cc4ac0fbc1f6e19
chrUn_KI270373v1    LN:1451 M5:8abd598da8e385b0f3c48f0b6bc9e35e
chrUn_KI270375v1    LN:2378 M5:efde09db3d9e65bcf7f2336634c86213
chrUn_KI270371v1    LN:2805 M5:a53c22fb527bcfd739209c23702ecce5
chrUn_KI270448v1    LN:7992 M5:8fe4b9bfb278721633fe864ba564f06b
chrUn_KI270521v1    LN:7642 M5:ef22c3c4cf8acf1418f3c46d082aeaaa

```

Below is Picard generated MD5

```
# hg38.dict generated from Picard
@SQ SN:chrUn_KI270374v1 LN:2656 M5:dbc92c9a92e558946e58b4909ec95dd5 UR:file:///mnt/scratch_ssd/Tian/Cansor/EMseq/13.PicardHSMetric/hg38.fa.gz
@SQ SN:chrUn_KI270372v1 LN:1650 M5:53a9d5e8fd28bce5da5efcfd9114dbf2 UR:file:///mnt/scratch_ssd/Tian/Cansor/EMseq/13.PicardHSMetric/hg38.fa.gz
@SQ SN:chrUn_KI270373v1 LN:1451 M5:b174fe53be245a840cd6324e39b88ced UR:file:///mnt/scratch_ssd/Tian/Cansor/EMseq/13.PicardHSMetric/hg38.fa.gz
@SQ SN:chrUn_KI270375v1 LN:2378 M5:d678250c97e9b94aa390fa46e70a6d83 UR:file:///mnt/scratch_ssd/Tian/Cansor/EMseq/13.PicardHSMetric/hg38.fa.gz
@SQ SN:chrUn_KI270371v1 LN:2805 M5:a0af3d778dfeb7963e8e6d84c0c54fba UR:file:///mnt/scratch_ssd/Tian/Cansor/EMseq/13.PicardHSMetric/hg38.fa.gz
@SQ SN:chrUn_KI270448v1 LN:7992 M5:0f40827c265cb813b6e723da6c9b926b UR:file:///mnt/scratch_ssd/Tian/Cansor/EMseq/13.PicardHSMetric/hg38.fa.gz

```

While the second MD5, generated by capital letter sequence is exactly the same as Picard. <b style="background-color: yellow">In other word, Picard use capital letters for MD5, but GemBS use lowercase letters for MD5</b>.

One intuitive solution is to manually write code to replace MD5 in hg38.dict with MD5 in `hg38.gemBS.contig_md5`. But this solution is not working, I tried. Because in the CollectHsMetric function, it will still re-calculate a MD5 for inputed reference. <b style="background-color: yellow">So the only way I found work, is to replace whole BAM file's header MD5, with Picard generated MD5.</b>.


## 3. @RD Error with missing SM

Then finally we can start to run CollectHSMetric, firstly some modification need to be done with GemBS produced BAM file. There is a head tag (@RG) have empty SM value. Like below:

```
# samtools view -H J080001.bam | less
@SQ     SN:chrUn_GL000216v2     LN:176608       M5:a44dc63aaa8abe4c33073ea553b58369
@SQ     SN:chrUn_GL000218v1     LN:161147       M5:6179aa420feaba9a16c86dca3d051b84
@SQ     SN:chrX LN:156040895    M5:6bdaf93397b486a58fd60b55aa2e21ca
@SQ     SN:chrY LN:57227415     M5:9bd609da53b41a50a724f2a0131ee9c1
@SQ     SN:chrY_KI270740v1_random       LN:37240        M5:1bccb354cf558625dbad9213faeee6ca
@RG     ID:J080001      SM:     BC:J080001      PU:J080001
@PG     ID:GEM  PN:gem-mapper   VN:v3.6.0-cnag-release  CL:/scratch1/Tian/Software/install/gemBS-rs/lib/gemBS/bin/gem-mapper --threads 20 -I index/hg38.BS.gem --i1 /scratch1/Tian/Cansor/EMseq/1.Fastp/JoinDataTmp/J080001_R1.fq.gz --i2 /scratch1/Tian/Cansor/EMseq/1.Fastp/JoinDataTmp/J080001_R2.fq.gz --paired-end-alignment --bisulfite-conversion inferred-C2T-G2A --report-file ./Compare3Set//mapping/J080001/J080001.json --sam-read-group-header @RG\\tID:J080001\\tSM:\\tBC:J080001\\tPU:J080001
@PG     ID:samtools     PN:samtools     PP:GEM  VN:1.14 CL:/scratch1/Tian/Software/install/gemBS-rs/lib/gemBS/bin/samtools sort -o ./Compare3Set//mapping/J080001/J080001.bam -T ./Compare3Set//mapping/J080001 --threads 20 --write-index -
@PG     ID:samtools.1   PN:samtools     PP:samtools     VN:1.10 CL:samtools view -H J080001.bam
@CO     TM:2022/1/20 02:58:36 CET       WD:/mnt/scratch_ssd/Tian/Cansor/EMseq/2.GemBS_Analysis  HN:slms-ci-beck-02      UN:tian

```

Previously I thought I can remove this line, but eventually I know I can't remove it. So the way is to add a value, like `None` to after SM. Make it like. In next step I will do it after I exported the header of BAM file.



## 4. Edit BAM Head

Now I need to find a way to edit the head of BAM file. Replace it with picard MD5. I manually replace the first one line of MD5 in a BAM file, seems it works.

Firstly I need to export the header of BAM as:

```
samtools view -H J080001.bam > header.sam

```

Then replace it with Picard generated MD5, rewrite a file as `header.new.sam`.

```
samMD5 <- read.csv("./header.sam", header=F)
picardMD5 <- read.csv("./hg38.dict", sep="\\t", header=F)

SQTag <- samMD5[2:nrow(picardMD5), 1]
SQTag <- sapply(SQTag, function(x) strsplit(x, split="M5")[[1]][1])
SQTag <- paste0(SQTag, picardMD5[-1, "V4"])

samMD5[2:nrow(picardMD5), 1] <- SQTag
write.table(samMD5, "header.new.sam", quote=F, row.names=F, col.names=F)
```

Now I use vim to edit the `header.new.sam` file the missing SM tag in @RG line, make it as:

```bash
@RG     ID:J080001      SM:None     BC:J080001      PU:J080001
```

Then, use samtools to reset header of the bam file, like below:

```
samtools reheader header.new.sam J080001.bam > J080001.newhead.bam

```

## 5. Run CollectHsMetric

Finally, I can run this function. It requires indexed reference, so use samtools to index the **unziped** reference.

```bash
samtools faidx hg38.fa
```

Then run comand on the modified `J080001.newhead.bam` file. 

```
java -jar picard.jar CollectHsMetrics -I J080001.newhead.bam -O J080001_hs_metrics.txt -R hg38.fa -BAIT_INTERVALS Twist.interval_list -TARGET_INTERVALS Twist.interval_list -PER_TARGET_COVERAGE J080001_per_target_coverage_hs_metrics.txt -COVERAGE_CAP 1000 -NEAR_DISTANCE 500

```
Finally it works!!!

---
## Summary

Honestly I am quite disappointed of this kind of work...spending days and nights on bug caused by version of software, detailed MD5 method .etc. Bioinformatic contains so many poor written software, without clear document, can't input each other easily. etc. All these "hard work" are useless, but necessarily need Bioinformaticians to take time and patient to debug them out...

What I am even more disappointed are tons of scientific research output based on these poorly written software.

I hope one day, I can open a company to provide **fast, convinient, high-quality, well-documented, and completely transparently explained software** for Bioinformatic research.